src/Features/Core/Portable/CustomMessageHandler/CustomMessageHandlerService.cs

```diff
@@ -0,0 +1,350 @@
+                        {
+                            _documentHandlers.Add(handler.Key, handler.Value);
+                        }
+
+                        extension.Assemblies[assemblyFileName] = new()
+                        {
+                            Handlers = messageHandlers.Keys.ToHashSet(),
+                            DocumentHandlers = messageDocumentHandlers.Keys.ToHashSet(),
+                        };
+
+                        return ValueTask.FromResult<RegisterHandlersResponse>(
+                            new(messageHandlers.Keys.ToArray(), messageDocumentHandlers.Keys.ToArray()));
+                    }
+                }
+                catch (Exception e)
+                {
+                    Logger.Log(
+                        FunctionId.CustomMessageHandlerService_HandleCustomMessageAsync,
+                        $"Error loading handlers from {assemblyPath}: {e}",
+                        LogLevel.Error);
+
+                    if (mustCleanupExtension)
+                    {
+                        extension.AssemblyLoadContext.Unload();
+                    }
+                    else
+                    {
+                        extension.Assemblies[assemblyFileName] = null;
+                    }
+                    throw;
+                }
+            }
+        }
+
+        Assembly? ResolveExtensionAssembly(AssemblyLoadContext context, AssemblyName assemblyName)
+        {
+            try
+            {
+                return _defaultLoadContext.LoadFromAssemblyName(assemblyName);
+            }
+            catch
+            {
+            }
+
+            var extensionAssemblyPath = Path.Combine(assemblyFolderPath, $"{assemblyName.Name}.dll");
+
+            // This will throw FileNotFoundException if the assembly is not found.
+            return context.LoadFromAssemblyPath(extensionAssemblyPath);
+        }
+    }
+
+    public async ValueTask<string> HandleCustomMessageAsync(
+        Solution solution,
+        string messageName,
+        string jsonMessage,
+        CancellationToken cancellationToken)
+    {
+#if DEBUG
+        System.Diagnostics.Debugger.Launch();
+#endif
+
+        ICustomMessageHandlerWrapper handler;
+        lock (_lockObject)
+        {
+            if (!_handlers.TryGetValue(messageName, out handler!))
+            {
+                throw new InvalidOperationException($"No handler found for message {messageName}.");
+            }
+        }
+
+        var message = JsonSerializer.Deserialize(jsonMessage, handler.MessageType);
+        var result = await handler.ExecuteAsync(message, solution, cancellationToken)
+            .ConfigureAwait(false);
+        var responseJson = JsonSerializer.Serialize(result, handler.ResponseType);
+        return responseJson;
+    }
+
+    public async ValueTask<string> HandleCustomDocumentMessageAsync(
+        Solution solution,
+        string messageName,
+        string jsonMessage,
+        DocumentId documentId,
```

Comment:
instead of Sol+DocId, our pattern woudl be to pass a Document instance instead directly instead of both.